pkg/bbgo/order_execution.go

```diff
@@ -321,30 +322,46 @@ func (c *BasicRiskController) ProcessOrders(
 
 type OrderCallback func(order types.Order)
 
+type batchOrderStatus struct {
+	Index int
+	Order *types.Order
+	Err   error
+}
+
 // BatchPlaceOrder
 func BatchPlaceOrder(
 	ctx context.Context, exchange types.Exchange, orderCallback OrderCallback, submitOrders ...types.SubmitOrder,
 ) (types.OrderSlice, []int, error) {
-	var createdOrders types.OrderSlice
-	var err error
+	g, _ := errgroup.WithContext(ctx)
 
-	var errIndexes []int
+	results := make([]batchOrderStatus, len(submitOrders))
 	for i, submitOrder := range submitOrders {
-		createdOrder, err2 := exchange.SubmitOrder(ctx, submitOrder)
-		if err2 != nil {
-			err = multierr.Append(err, err2)
-			errIndexes = append(errIndexes, i)
-		}
+		i, submitOrder := i, submitOrder
+		g.Go(func() error {
+			order, err := exchange.SubmitOrder(ctx, submitOrder)
+			results[i] = batchOrderStatus{i, order, err}
+			return nil
+		})
+	}
 
-		// if createdOrder is not nil and has a valid OrderID, we should add it to the createdOrders
-		if createdOrder != nil && createdOrder.OrderID > 0 {
-			createdOrder.Tag = submitOrder.Tag
+	_ = g.Wait()
 
+	var (
+		createdOrders types.OrderSlice
+		errIndexes    []int
+		err           error
+	)
+
+	for _, result := range results {
+		if result.Err != nil {
+			err = multierr.Append(err, result.Err)
+			errIndexes = append(errIndexes, result.Index)
+		} else if result.Order != nil {
+			result.Order.Tag = submitOrders[result.Index].Tag
 			if orderCallback != nil {
-				orderCallback(*createdOrder)
+				orderCallback(*result.Order)
```

Comment:
you need to call orderCallback immediately once you submitted the order, this is too far from the call.